src/main/java/org/jenkins/plugins/lockableresources/LockStep.java

```diff
@@ -10,26 +10,50 @@
 
 import hudson.Extension;
 import hudson.model.AutoCompletionCandidates;
+import hudson.util.FormValidation;
+import hudson.Util;
+
+import edu.umd.cs.findbugs.annotations.Nullable;
+import edu.umd.cs.findbugs.annotations.CheckForNull;
 
 public class LockStep extends AbstractStepImpl implements Serializable {
 
-	public final String resource;
+	@CheckForNull
+	public String resource = null;
+
+	@CheckForNull
+	public String label = null;
+
+	public int quantity = 0;
 
 	public boolean inversePrecedence = false;
 
+	// it should be LockStep() - without params. But keeping this for backward compatibility
+	// so `lock('resource1')` still works and `lock(label: 'label1', quantity: 3)` works too (resource is not required)
 	@DataBoundConstructor
 	public LockStep(String resource) {
-		if (resource == null || resource.isEmpty()) {
-			throw new IllegalArgumentException("must specify resource");
+		if (resource != null && !resource.isEmpty()) {
+			this.resource = resource;
 		}
-		this.resource = resource;
 	}
 
 	@DataBoundSetter
 	public void setInversePrecedence(boolean inversePrecedence) {
 		this.inversePrecedence = inversePrecedence;
 	}
 
+	@DataBoundSetter
+	public void setLabel(String label) {
+		if (!label.isEmpty()) {
```

Comment:
It's never going to be null because `setLabel` is only called if there is an actual value for label in the step config. But yeah, let's add the check anyway.